Tutorial on creating a bare metal application (#1305)
ARMmbed/Handbook: docs/bare_metal/using_bare_metal.md

# Using the bare metal profile

This guide shows how to create a bare metal profile application:
1. By default, when you build an application binary, the build tool uses the full profile. To use the bare metal profile, you need to set up your application to override this default behaviour.
1. Bare metal has a minimal set of APIs. You can add additional ones [from the list of supported API](bare_metal.md#features).

## Creating a bare metal application

> 1. Create a new Mbed application and enter its directory:

Reviewer: I would introduce the bare metal application that we are building (blinky using EventQueue) by describing what the application does and showing its `main.cpp`.

Author: `main.cpp` is quoted below (if the doc compiler inlines the link). Do you suggest showing it first?

    ```
    mbed new example_app && cd example_app
    ```

    This contains an empty application with Mbed OS (`mbed-os/`) pre-fetched.

1. Open `mbed_app.json` and replace it with the following content:

    ```json
    {
        "requires": ["bare-metal"],
        "target_overrides": {
            "*": {
                "target.c_lib": "small"
            }
        }
    }
    ```
    <span class="tips">**Tip:** `"target.c_lib": "small"` enables an optimised version of the C library with lower memory footprint. For more details, see [Using small C libraries in Mbed OS bare metal](c_small_libs.md)</span>.

    Bare metal has a minimal set of default APIs - those that are always available to a bare metal application. You can add other supported APIs if you need the features they enable.

>     For a list of default and supported APIs, [please see our full API list](bare_metal.md#features).

Reviewer: Irit should update to the correct link.

Author: Cool. I've seen both md and html before and am a bit confused.

> 1. To add an API - in this example, the `EventQueue` class:

Reviewer: ```suggestion
1. Add dependencies - in this example, we need to add the library that contains the `EventQueue` class:
```

>     1. Locate the library you want to use in `mbed-os/`.

Reviewer: Suggestion: Locate the API and the library where it is declared in `mbed-os/`

    1. In the library folder, find the library's name in `mbed_lib.json`. You will need it for the next step.

        For example: `events/mbed_lib.json`:
        ```json
        {
            "name": "events",
            "config": {
                "present": 1,
                ...
            }
        }
        ```
        To continue, go back to the application's root directory.

    1. Open `mbed_app.json` again, and add the library to the `"requires"` field:

        ```json
        {
            "requires": [
                "bare-metal",

>                 "events"

Reviewer: Minor comment: I prefer to see all the required libraries in one line. It helps when using grep.

Author: I'll change it to be in one line since there're only two items.

            ]
        }
        ```

1. Write the application using the libraries you added.

>     [![View code](https://www.mbed.com/embed/?url=https://github.com/armmbed/mbed-os-example-baremetal-eventqueue-blinky/)](https://github.com/ARMmbed/mbed-os-example-baremetal-eventqueue-blinky/blob/master/main.cpp)

Reviewer: We will have to create a code snippet in https://github.com/ARMmbed/mbed-os-examples-docs_only

Author: In `baremetal_example.md` which I deleted, the blinky baremetal example repository's `main.cpp` was added in the way and somehow compiled into a code block the public doc. I was expecting this to be how it works?

Reviewer 2: Yes, we can include a code block from any repository.

Author: @evedon pointed out the link I used is not an official example and we want to avoid it in documentations. I've found another very simple EventQueue snippet:
https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/APIs_RTOS/EventQueue_ex_2/main.cpp
Will use this one instead.

# Using the bare metal profile

This guide shows how to create a bare metal profile application:

> 1. By default, when you build an application binary, the build tool uses the full profile. To use the bare metal profile, you need to set up your application to override this default behaviour.

Reviewer 2: ```suggestion
1. Set the profile: By default, the build tool uses the full profile for all application builds. To use the bare metal profile, set up your application to override this default behaviour.
```

> 1. Bare metal has a minimal set of APIs. You can add additional ones [from the list of supported API](bare_metal.md#features).

Reviewer 2: ```suggestion
1. By default, the bare metal profile uses a minimal set of APIs. You can add additional ones [from the list of supported APIs](../bare-metal/index.html#features) if your application needs them.
```

## Creating a bare metal application

> To demonstrate how to create a bare metal application, we use an example that prints text in regular intervals using the `EventQueue` class:

Reviewer 2: ```suggestion
To demonstrate how to create a bare metal application, here is an example that prints text at regular intervals using the `EventQueue` class:
```

> [![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/APIs_RTOS/EventQueue_ex_2/)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/APIs_RTOS/EventQueue_ex_2/main.cpp)

Reviewer 2: I need to double-check that this transcluded correctly.

> The steps are as follows:

Reviewer 2: ```suggestion
To create the application:
```

> 1. Create a new Mbed application and enter its directory:

Reviewer 2: ```suggestion
1. Create a new Mbed OS application and navigate to its directory:
```

    ```
    mbed new example_app && cd example_app
    ```

>     This contains an empty application with Mbed OS (`mbed-os/`) fetched.

Reviewer 2: ```suggestion
    The directory contains the full Mbed OS library (`mbed-os/`) and no application files.
```

> 1. Create a `main.cpp` containing the `EventQueue` example code snippet above.

Reviewer 2: ```suggestion
1. Create a `main.cpp` file containing the `EventQueue` code snippet above.
```

> 1. Open `mbed_app.json` and replace it with the following content:

Reviewer 2: ```suggestion
1. Open `mbed_app.json` (in the root of the application) and replace it with the following content:
```

    ```json
    {
        "requires": ["bare-metal"],
        "target_overrides": {
            "*": {
                "target.c_lib": "small"
            }
        }
    }

>     ```

Reviewer 2: ```suggestion
    ```

    The file specifies which profile to use (`"requires": ["bare-metal"]`) and which C library to use (`"target.c_lib": "small"`).
```

>     <span class="tips">**Tip:** `"target.c_lib": "small"` enables an optimised version of the C library with lower memory footprint. For more details, see [Using small C libraries in Mbed OS bare metal](c_small_libs.md)</span>.

Reviewer 2: ```suggestion
    In this example, we're using `"target.c_lib": "small"` (small C library). This means your application will use an optimised version of the C library with lower memory footprint. For more details, see [Using small C libraries in Mbed OS bare metal](../bare-metal/using-small-c-libraries.html).
```

> 

Reviewer 2: Add:

You now have application code and a bare metal profile with the default APIs. However, this example uses APIs that are not part of the default bare metal profile - you need to manually add support for those APIs to the application.

## Adding APIs

Author: This adds a new section, so I'll adjust sections for steps above and below accordingly.

    Bare metal has a minimal set of default APIs - those that are always available to a bare metal application. You can add other supported APIs if you need the features they enable.

>     For a list of default and supported APIs, [please see our full API list](bare_metal.md#features).

Reviewer 2: ```suggestion
    For a list of default and supported APIs, [please see our full API list](../apis/index.html).
```

> 1. Add dependencies - in this example, we need to add the library that contains the `EventQueue` class:

Reviewer 2: ```suggestion
1. This example depends on the `EventQueue` class, so you need to add the library that contains that class:
```

>     1. Locate the API and the library where it is declared in `mbed-os/`.

Reviewer 2: ```suggestion
    1. In `mbed-os/`, locate the API and the library in which it is declared.
```

>     1. In the library folder, find the library's name in `mbed_lib.json`. You will need it for the next step.

Reviewer 2: ```suggestion
    1. In the library folder, open `mbed_lib.json` and find the library's name. You will need it for the next step.
```

        For example: `mbed-os/events/mbed_lib.json`:
        ```json
        {
            "name": "events",
            "config": {
                "present": 1,
                ...
            }
        }
        ```
        To continue, go back to the application's root directory.

    1. Open `mbed_app.json` again, and add the library to the `"requires"` field:

        ```json
        {
            "requires": ["bare-metal", "events"],
            "target_overrides": {
                "*": {
                    "target.c_lib": "small"
                }
            }
        }
        ```

> 1. With a supported board connected to your computer, compile and run your application:

Reviewer 2: ```suggestion
1. Connect a supported board to your computer, and compile and run your application:
```

    ```
    mbed compile -t <TOOLCHAIN> -m <TARGET> --flash --sterm
    ```

>     This opens a serial terminal once the example is compiled and flashed. The following output is expected:

Reviewer 2: ```suggestion
    When the example is flashed, a serial terminal opens (because of `--sterm`). You should get the following output:
```

    ```
    called immediately
    called every 1 seconds
    called in 2 seconds
    called every 1 seconds
    called every 1 seconds
    called every 1 seconds
    ```

>     To exit the serial terminal, press Ctrl + C.

Reviewer 2: What if I don't get that output? What does it mean and what should I do?

Author: There can be different causes of not getting the expected output - issues with serial communication, target drivers, the application itself, etc. In my opinion it may not be feasible to have a troubleshooting guide wherever some output is quoted.

In https://os.mbed.com/docs/mbed-os/v5.15/tutorials/the-eventqueue-api.html for example, it just says "This is the output of the above program on an FRDM-K64F board." So I'll replace "You should get the following output" which seems to hint users they may or may not get it.
